src/OVAL/oval_schema_version.c

```diff
@@ -0,0 +1,157 @@
+	const char *error;
+	int erroffset;
+	pcre *re = pcre_compile(pattern, 0, &error, &erroffset, NULL);
+	if (re == NULL) {
+		dE("Regular expression compilation failed with %s\n", pattern);
+		return version;
+	}
+	const int ovector_size = 30; // must be a multiple of 30
+	int ovector[ovector_size];
+	int rc = pcre_exec(re, NULL, ver_str, strlen(ver_str), 0, 0, ovector, ovector_size);
+	pcre_free(re);
+	if (rc < 0) {
+		dE("Regular expression %s did not match string %s\n", pattern, ver_str);
+		return version;
+	}
+	for (int i = 1; i < rc; i++) {
+		const char *substring = ver_str + ovector[2 * i];
+		size_t substring_length = ovector[2 * i + 1] - ovector[2 * i];
+		version.component[i - 1] = _parse_int(substring, substring_length);
+	}
+#elif defined USE_REGEX_POSIX
+	const char *pattern = "([0-9]+)\\.([0-9]+)(\\.([0-9]+))?(:([0-9]+)\\.([0-9]+)(\\.([0-9]+))?)?";
+	int groups[OVAL_SCHEMA_VERSION_COMPONENTS_COUNT] = {1, 2, 4, 6, 7, 9};
+	regex_t re;
+	if (regcomp(&re, pattern, REG_EXTENDED) != 0) {
+		dE("Regular expression compilation failed with %s\n", pattern);
+		return version;
+	}
+	size_t max_matches = 10;
+	regmatch_t matches[max_matches];
+	int match = regexec(&re, ver_str, max_matches, matches, 0);
+	regfree(&re);
+	if (match != 0) {
+		dE("Regular expression %s did not match string %s\n", pattern, ver_str);
+		return version;
+	}
+	for (int i = 0; i < OVAL_SCHEMA_VERSION_COMPONENTS_COUNT; i++) {
+		int group = groups[i];
+		if (matches[group].rm_so >= 0) {
+			const char *substring = ver_str + matches[group].rm_so;
+			size_t substring_length = matches[group].rm_eo - matches[group].rm_so;
+			version.component[i] = _parse_int(substring, substring_length);
+		}
+	}
+#endif
+	return version;
+}
+
+const char *oval_schema_version_to_cstr(oval_schema_version_t version)
+{
+	size_t buf_len = 32;
+	char buf[buf_len];
+	const char *format;
+	if (version.component[OVAL_SCHEMA_VERSION_PLATFORM_MAJOR]) {
+		if (version.component[OVAL_SCHEMA_VERSION_CORE_UPDATE] &&
+		version.component[OVAL_SCHEMA_VERSION_PLATFORM_UPDATE]) {
+			format = "%1$d.%2$d.%3$d:%4$d.%5$d.%6$d";
+		} else if (version.component[OVAL_SCHEMA_VERSION_CORE_UPDATE]) {
+			format = "%1$d.%2$d.%3$d:%4$d.%5$d";
+		} else if (version.component[OVAL_SCHEMA_VERSION_PLATFORM_UPDATE]) {
+			format = "%1$d.%2$d:%4$d.%5$d.%6$d";
+		} else {
+			format = "%1$d.%2$d:%4$d.%5$d";
+		}
+	} else if (version.component[OVAL_SCHEMA_VERSION_CORE_UPDATE]) {
+		format = "%1$d.%2$d.%3$d";
+	} else {
+		format = "%1$d.%2$d";
+	}
+	snprintf(buf, buf_len, format,
+		version.component[OVAL_SCHEMA_VERSION_CORE_MAJOR],
+		version.component[OVAL_SCHEMA_VERSION_CORE_MINOR],
+		version.component[OVAL_SCHEMA_VERSION_CORE_UPDATE],
+		version.component[OVAL_SCHEMA_VERSION_PLATFORM_MAJOR],
+		version.component[OVAL_SCHEMA_VERSION_PLATFORM_MINOR],
+		version.component[OVAL_SCHEMA_VERSION_PLATFORM_UPDATE]);
+	return oscap_strdup(buf);
```

Comment:
I don't understand the reasoning behind the strdup here. We create a buf on the stack, we strdup it on the heap and destroy it on the stack upon return. Why not allocate it on the heap in the first place and just return it?

Maybe I am missing something.